RxSwift/Subjects/AsyncSubject.swift

```diff
@@ -0,0 +1,138 @@
+//
+//  AsyncSubject.swift
+//  RxSwift
+//
+//  Created by Victor Galán on 07/01/2017.
+//  Copyright © 2017 Krunoslav Zaher. All rights reserved.
+//
+
+import Foundation
+
+/// Represents a value that changes over time.
+///
+/// Observers can subscribe to the subject to receive the last (or initial) value and all subsequent notifications.
+public final class AsyncSubject<Element>
+	: Observable<Element>
+	, SubjectType
+	, ObserverType
+	, SynchronizedUnsubscribeType
+	, Disposable {
+	public typealias SubjectObserverType = AsyncSubject<Element>
+	typealias DisposeKey = Bag<AnyObserver<Element>>.KeyType
+
+	/// Indicates whether the subject has any observers
+	public var hasObservers: Bool {
+		_lock.lock(); defer { _lock.unlock() }
+		return _observers.count > 0
+	}
+
+	let _lock = NSRecursiveLock()
+
+	// state
+	private var _isDisposed = false
+	private var _observers = Bag<AnyObserver<Element>>()
+	private var _stoppedEvent: Event<Element>?
+	private var _stopped = false
+	private var _lastValue: Element?
+
+	/// Indicates whether the subject has been disposed.
+	public var isDisposed: Bool {
+		return _isDisposed
+	}
+
+	/// Creates a subject.
+	public override init() {
+		super.init()
+	}
+
+	/// Notifies all subscribed observers about next event.
+	///
+	/// - parameter event: Event to send to the observers.
+	public func on(_ event: Event<E>) {
+		_synchronized_on(event)
+	}
+
+	func _synchronized_on(_ event: Event<E>) {
+		_lock.lock(); defer { _lock.unlock() }
+		if  _isDisposed {
+			return
+		}
+
+		switch event {
+		case .next(let value):
+			if !_stopped {
+				_lastValue = value
+			}
+		case .error, .completed:
+			_stoppedEvent = event
+			_stopped = true
+		}
+
+		if let stoppedEvent = _stoppedEvent {
```

Comment:
I think cases for error and completed should be separated
```
case .error
case .completed:
```

so there is no need for additional `if let` check.

I understand that your goal was to remove duplication of logic but in this case I think the code will have less conditionals and execute faster because this check will be performed for each event vs once per terminal event.

It maybe looks like a microoptimization, but last time I've checked `if let` construct was quite slow. That's why there is `_stopped` flag also.